java/arcs/core/analysis/RecipeGraphFixpointIterator.kt

```diff
@@ -0,0 +1,107 @@
+ */
+
+package arcs.core.analysis
+
+import arcs.core.data.HandleConnectionSpec
+import arcs.core.data.Recipe
+
+/** An abstract class that implements dataflow analysis over abstract values of type [V]. */
+abstract class RecipeGraphFixpointIterator<V : AbstractValue<V>> {
+    /**
+     * Abstract state at the node, i.e., the join of the abstract values along the input edges.
+     *
+     * Absence of an entry for a node denotes BOTTOM.
+     */
+    private var nodeValues = mutableMapOf<RecipeGraph.Node, V>()
+
+    /** State transformer for a [Recipe.Handle] node. */
+    open fun transform(handle: Recipe.Handle, input: V): V = input
+
+    /** State transformer for a [Recipe.Particle] node. */
+    open fun transform(particle: Recipe.Particle, input: V): V = input
+
+    /** State transformer for a [Recipe.Handle] -> [Recipe.Particle] edge. */
+    open fun transform(
+        handle: Recipe.Handle,
+        particle: Recipe.Particle,
+        spec: HandleConnectionSpec,
+        input: V
+    ): V = input
+
+    /** State transformer for a [Recipe.Particle] -> [Recipe.Handle] edge. */
+    open fun transform(
+        particle: Recipe.Particle,
+        handle: Recipe.Handle,
+        spec: HandleConnectionSpec,
+        input: V
+    ): V = input
+
+    /**
+     * Returns the initial values for the nodes for starting a fixpoint iteration.
+     *
+     * The nodes in the returned map are used to initialize the worklist for the fixpoint iteration.
+     */
+    abstract fun getInitialValues(graph: RecipeGraph): Map<RecipeGraph.Node, V>
+
+    /** Returns the value for the given particle. Returns null if the value is BOTTOM. */
+    fun getValue(particle: Recipe.Particle): V? = nodeValues[RecipeGraph.Node.Particle(particle)]
+
+    /** Returns the value for the given handle. Returns null if the value is BOTTOM. */
+    fun getValue(handle: Recipe.Handle): V? = nodeValues[RecipeGraph.Node.Handle(handle)]
+
+    fun computeFixpoint(graph: RecipeGraph) {
+        // TODO(bgogul): If there are identical particles in the recipe, the results will be messed
+        // up. Need to fix the issue.
+        nodeValues = getInitialValues(graph).toMutableMap()
+        val worklist = nodeValues.keys.toMutableSet()
+        while (worklist.isNotEmpty()) {
+            // Pick and remove an element from the worklist.
+            val current = worklist.first()
+            worklist.remove(current)
+            val input = nodeValues[current]
+            // Treating as BOTTOM now. See `successors.forEach` below as well.
+            if (input == null) continue
+            val output = transform(current, input)
+            current.successors.forEach { (succ, spec) ->
+                val edgeValue = transform(current, succ, spec, output)
+                val oldValue = nodeValues[succ]
+                val newValue = oldValue?.join(edgeValue) ?: edgeValue
+                val changed = oldValue?.isEquivalentTo(newValue)?.not() ?: true
```

Comment:
This line is pretty hard to understand. Could you do something like: `oldValue == null || !(oldValue isEquivalentTo newValue)`? (Please double check the logic is the same)

Reply:
Simplified this quite a bit now by making bottoms explicit.